### PR TITLE
placeFetcher extend radius

### LINKED ISSUE #28: change radius in PlacesFetcher to grow depending in number of results

### PR DESCRIPTION
**Pull request by tbarnahor**
The purpose of this PR is to extend the fetching radius by 5km each time (up to 20 km) until at least 10 results are found.

### CODE REVIEW

File: project/src/main/java/com/google/sps/data/PlacesFetcher.java [hunk 1]
```diff
@@ -37,9 +37,14 @@
      */
     private static final PlaceType TYPE = PlaceType.RESTAURANT;
 
-    /** In this radius around "LOCATION" places will be searched. */
-    private static final int SEARCH_RADIUS = 5000;
-    // TODO(M1): check at least 10 results, and if less extend radius
+    /** In this radius around "LOCATION" places will be searched initially. */
+    private static final int INIT_SEARCH_RADIUS = 5000;
```
Comment (karenav) [inline project/src/main/java/com/google/sps/data/PlacesFetcher.java]: It's pretty obvious for us, but maybe it should be pointed out that it's in meters?

File: project/src/main/java/com/google/sps/data/PlacesFetcher.java [hunk 2]
```diff
@@ -55,20 +60,31 @@
...
-            query.openNow(preferences.openNow());
-        }
-        try {
-            return createPlacesList(getPlacesSearchResults(query));
-        } catch (ApiException | InterruptedException | IOException e) {
-            throw new FetcherException("Couldn't fetch places from Places API", e);
+            request.openNow(preferences.openNow());
         }
+        return request;
     }
```
Comment (karenav) [inline project/src/main/java/com/google/sps/data/PlacesFetcher.java:88]: The whole change looks really good!!
It took me a little while to understand what was going on, and I think it might have been more readable for me if MAX_RADIUS_MULT was called something else, like MAX_NUM_ITERATIONS_OF_RADIUS_CHANGE (or something else in that field).

Comment (tbarnahor) [reply project/src/main/java/com/google/sps/data/PlacesFetcher.java:88]: Right! changed it

File: project/src/main/java/com/google/sps/data/PlacesFetcher.java (hunk 1 above)
Comment (tbarnahor) [reply project/src/main/java/com/google/sps/data/PlacesFetcher.java]: Right! changed it

File: project/src/main/java/com/google/sps/data/PlacesFetcher.java [hunk 3]
```diff
@@ -37,9 +37,14 @@
      */
     private static final PlaceType TYPE = PlaceType.RESTAURANT;
 
-    /** In this radius around "LOCATION" places will be searched. */
-    private static final int SEARCH_RADIUS = 5000;
-    // TODO(M1): check at least 10 results, and if less extend radius
+    /** In this radius (in meters) around "LOCATION" places will be searched initially.*/
```
Comment (arielz-g) [inline project/src/main/java/com/google/sps/data/PlacesFetcher.java]: Use // commenting style for privates, /** is for javadoc purposes (same for other fields).
Also, perhaps name this as "INIT_SEARCH_RADIUS_MR" to suggest the unit.

nit for phrasing: The initial search radius in meters.

File: project/src/main/java/com/google/sps/data/PlacesFetcher.java [hunk 4]
```diff
@@ -55,20 +60,32 @@
      *     for places or for places details
      */
     public ImmutableList<Place> fetch(UserPreferences preferences) throws FetcherException {
-        TextSearchRequest query =
-            PlacesApi.textSearchQuery(
-                CONTEXT, createCuisinesQuery(preferences.cuisines()), preferences.location())
-                .radius(SEARCH_RADIUS)
+        PlacesSearchResult[] placesSearchResult;
+        int radiusMult = 1;
```
Comment (arielz-g) [inline project/src/main/java/com/google/sps/data/PlacesFetcher.java]: attemptsCounter?

Comment (arielz-g) [inline project/src/main/java/com/google/sps/data/PlacesFetcher.java]: Initialize this with 0 and change below to < instead of <=

File: project/src/test/java/com/google/sps/data/PlacesFetcherTest.java [hunk 5]
```diff
@@ -209,4 +209,18 @@ public void fetch_PlaceDetailsQueryFails_throwsFetcherException() throws Excepti
     assertTrue(thrown.getCause() instanceof IOException);
     assertTrue(thrown.getMessage().contains("place details"));
   }
+
+  @Test
+  public void fetch_resultsOnlyAfterRadiusExtension_returnsListOfPlaces() throws Exception {
```
Comment (arielz-g) [inline project/src/test/java/com/google/sps/data/PlacesFetcherTest.java:220]: Add a test that retrieves nothing after max number of extensions was reached to handle the edge case.

File: project/src/main/java/com/google/sps/data/PlacesFetcher.java (hunk 3 above)
Comment (tbarnahor) [reply project/src/main/java/com/google/sps/data/PlacesFetcher.java]: Changed it.

File: project/src/main/java/com/google/sps/data/PlacesFetcher.java (hunk 4 above)
Comment (tbarnahor) [reply project/src/main/java/com/google/sps/data/PlacesFetcher.java]: I wanted it to be the factor the initial radius is multiplied by but indeed it is not clear, so I changed it like you suggested because one of his jobs it to also count attempts.

Comment (tbarnahor) [reply project/src/main/java/com/google/sps/data/PlacesFetcher.java]: I initialized it to 1 because it multiplies the initial search radius (which if first multiplied by 1). I changed it to be initiated to 0 and updated in the begining of the attempt so I hope now it's clearer.

And you are right about < , changed it.

File: project/src/test/java/com/google/sps/data/PlacesFetcherTest.java (hunk 5 above)
Comment (tbarnahor) [reply project/src/test/java/com/google/sps/data/PlacesFetcherTest.java:220]: The fetch_zeroSearchResults_returnsEmptyList test checks this corner case. In this test the getPlacesSearchResults always returns nothing so this test ends only when max number of extensions were reached and recieves nothing.

**Review by karenav — APPROVED** [review]

File: project/src/main/java/com/google/sps/data/PlacesFetcher.java [hunk 6]
```diff
@@ -31,17 +31,21 @@
...
+    // rather as "RESTAURANT" this is the most suitable type to search for.
     private static final PlaceType TYPE = PlaceType.RESTAURANT;
 
-    /** In this radius around "LOCATION" places will be searched. */
-    private static final int SEARCH_RADIUS = 5000;
-    // TODO(M1): check at least 10 results, and if less extend radius
+    // The initial search radius in meters.
+    private static final int INIT_SEARCH_RADIUS_M = 5000;
+
+    // The minimum number of results to be fetched.
```
Comment (arielz-g) [inline project/src/main/java/com/google/sps/data/PlacesFetcher.java]: nit: minimal

File: project/src/main/java/com/google/sps/data/PlacesFetcher.java [hunk 7]
```diff
@@ -31,17 +31,21 @@
...
+    private static final int MIN_NUM_OF_RESULTS = 10;
+
+    // The maximum number of times the search radius will be extended.
```
Comment (arielz-g) [inline project/src/main/java/com/google/sps/data/PlacesFetcher.java]: nit: maximal

File: project/src/test/java/com/google/sps/data/PlacesFetcherTest.java (hunk 5 above)
Comment (arielz-g) [reply project/src/test/java/com/google/sps/data/PlacesFetcherTest.java:220]: I see, in that case please add a verification in the test that the mock was called MAX_NUM_OF_RADIUS_EXTENSIONS. You can use mockito's verify method, something like:
Mockito.verify(spiedFetcher, times(MAX_NUM_OF_RADIUS_EXTENSIONS)).getPlacesSearchResults()

see https://javapointers.com/java/unit-test/use-verify-in-mockito/ for more info.

**Review by arielz-g — APPROVED** [review]
Approving with minor fixes, LMK if something's unclear:)
